clientlibrary/worker/worker.go

```diff
@@ -204,19 +226,97 @@ func (w *Worker) initialize() error {
 	return nil
 }
 
+// fetchConsumerARNWithRetry tries to fetch consumer ARN. Retries 10 times with exponential backoff in case of an error
+func (w *Worker) fetchConsumerARNWithRetry() (string, error) {
+	for retry := 0; ; retry++ {
+		consumerARN, err := w.fetchConsumerARN()
+		if err == nil {
+			return consumerARN, nil
+		}
+		if retry < 10 {
+			sleepDuration := time.Duration(math.Exp2(float64(retry))*100) * time.Millisecond
+			w.kclConfig.Logger.Errorf("Could not get consumer ARN: %v, retrying after: %s", err, sleepDuration)
+			time.Sleep(sleepDuration)
+			continue
+		}
+		return consumerARN, err
+	}
+}
+
+// fetchConsumerARN gets enhanced fan-out consumerARN.
+// Registers enhanced fan-out consumer if the consumer is not found
+func (w *Worker) fetchConsumerARN() (string, error) {
+	log := w.kclConfig.Logger
+	log.Debugf("Fetching stream consumer ARN")
+	streamDescription, err := w.kc.DescribeStream(&kinesis.DescribeStreamInput{
+		StreamName: &w.kclConfig.StreamName,
+	})
+	if err != nil {
+		log.Errorf("Could not describe stream: %v", err)
+		return "", err
+	}
+	streamConsumerDescription, err := w.kc.DescribeStreamConsumer(&kinesis.DescribeStreamConsumerInput{
+		ConsumerName: &w.kclConfig.EnhancedFanOutConsumerName,
+		StreamARN:    streamDescription.StreamDescription.StreamARN,
+	})
+	if err == nil {
+		log.Infof("Enhanced fan-out consumer found, consumer status: %s", *streamConsumerDescription.ConsumerDescription.ConsumerStatus)
+		if *streamConsumerDescription.ConsumerDescription.ConsumerStatus != kinesis.ConsumerStatusActive {
+			return "", fmt.Errorf("consumer is not in active status yet, current status: %s", *streamConsumerDescription.ConsumerDescription.ConsumerStatus)
+		}
+		return *streamConsumerDescription.ConsumerDescription.ConsumerARN, nil
+	}
+	if awsErrCode(err) == kinesis.ErrCodeResourceNotFoundException {
+		log.Infof("Enhanced fan-out consumer not found, registering new consumer with name: %s", w.kclConfig.EnhancedFanOutConsumerName)
+		out, err := w.kc.RegisterStreamConsumer(&kinesis.RegisterStreamConsumerInput{
+			ConsumerName: &w.kclConfig.EnhancedFanOutConsumerName,
+			StreamARN:    streamDescription.StreamDescription.StreamARN,
+		})
+		if err != nil {
```

Comment:
May need to add some retries, like:
https://github.com/awslabs/amazon-kinesis-client/blob/cce05bf4fd03eed007bf6c51203f3cccc189c92a/amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/fanout/FanOutConfig.java#L76